modules/puppet_proxy/puppet_cache.rb

```diff
@@ -1,100 +1,44 @@
-            logger.debug("Using cached class #{puppetmodule}")
-            modulemanifest = lcache[puppetmodule]['manifest']
-          else
-            changed = true
-            logger.info("Scanning class #{puppetmodule}")
-            modulemanifest = Dir.glob("#{path}/manifests/**/*.pp").map do |filename|
-              scanner.scan_manifest File.read(filename), filename
-            end
-
-            lcache[puppetmodule]['timestamp']= Time.new.to_i
-            lcache[puppetmodule]['manifest'] = modulemanifest
-          end
-          modulemanifest
-        end.compact.flatten
-
-        if changed
-          logger.info("Cache file need to be updated for #{environment}: #{directory}")
-          # Clean obsolete cache modules
-          oldlength = lcache.length
-          lcache.delete_if { |key, value| !seenmodules.include?(key) }
-          logger.info("Cleaning #{oldlength - lcache.length } modules from cache") if oldlength - lcache.length > 0
-
-          cache[directory] = lcache
-          write_to_cache(cache, environment)
-          logger.info("Cache file updated for #{environment}: #{directory}")
-        end
-
-        manifest
-      end
-
-      def read_from_cache environment
-        cachefile = File.expand_path("cache_#{environment}.json", Proxy::Puppet::Plugin.settings.cache_location)
+    def initialize(class_scanner, classes_store, timestamps_store)
+      @classes_cache = classes_store
+      @timestamps = timestamps_store
+      @class_scanner = class_scanner
+      @mutex = Mutex.new
+    end
 
-        if File.exist?(cachefile)
-          as_hash = JSON.load(File.new(cachefile))
+    def scan_directory directory, environment
+      logger.debug("Running scan_directory on #{environment}: #{directory}")
 
-          if as_hash.delete('version') != ::Proxy::Puppet::Plugin.version
-            logger.info("Mismatching puppet cache version, discarding.")
-            return {}
-          end
+      @mutex.synchronize do
+        tmp_classes = MemoryStore.new
+        tmp_timestamps = MemoryStore.new
 
-          # ran into issues with deserialisation in older versions of json gem (ruby 1.8.7 + json 1.5.5),
-          # and opted for the deserialisation approach below.
-          # TODO: switch back to json library based deserialisation by implementing PuppetClass.json_create
-          # when support for Ruby 1.8.7 has been dropped.
-          as_hash.values.each do |dir|
-            dir.values.each do |puppet_module|
-              puppet_module['manifest'] =
-                puppet_module['manifest'].flatten.map { |klass| [Proxy::Puppet::PuppetClass.from_hash(klass)] }
+        Dir.glob("#{directory}/*").map do |path|
+          puppetmodule = File.basename(path)
+          Dir.glob("#{path}/manifests/**/*.pp").map do |filename|
+            if @timestamps[directory, filename] && (File.mtime(filename).to_i <= @timestamps[directory, filename])
```

Comment:
Also disregard please, thought this was the scan_manifest part of the conditional.